src/heartbeats.cpp

```diff
@@ -494,52 +574,65 @@ void MegaBackupMonitor::onSyncStateChanged(MegaApi* /*api*/, MegaSync *sync)
-    {
-        syncTag = mTSPair->second;
+        mHeartBeatedSyncs.erase(hBPair); //This is speculative: could be moved to backupremove_result
+        // in case we wanted to handle possible failing cases.
+        mPendingSyncUpdates.erase(sync->getTag()); // remove any pending update: also speculative.
     }
+}
+#endif
 
-    if (!syncTag) //first time
+std::shared_ptr<HeartBeatTransferProgressedInfo> MegaBackupMonitor::getHeartBeatBackupInfoByTransfer(MegaTransfer *transfer)
+{
+#ifdef ENABLE_SYNC
+    if (transfer->isSyncTransfer())
     {
-        Node *n = mClient->nodebyhandle(transfer->getType() == MegaTransfer::TYPE_UPLOAD ? transfer->getParentHandle() : transfer->getNodeHandle());
-        while (n)
+        int syncTag = 0;
+
+        // use map to get the syncTag directly if there was one
+        auto mTSPair = mTransferToSyncMap.find(transfer->getTag());
+        if (mTSPair != mTransferToSyncMap.end())
         {
-            if (n && n->localnode && n->localnode->sync)
-            {
-                syncTag = n->localnode->sync->tag;
-                mTransferToSyncMap[transfer->getTag()] = syncTag;
-                break;
-            }
-            LOG_warn << "Heartbeat could not get sync tag direclty from transfer handle. Going up";
-            n = n->parent;
+            syncTag = mTSPair->second;
         }
-    }
 
-    if (syncTag)
-    {
-        auto hBPair = mHeartBeatedSyncs.find(syncTag);
-        if (hBPair != mHeartBeatedSyncs.end()) //transfer info arrived before having created the HeartBeatSyncInfo object
+        if (!syncTag) //first time
         {
-            return hBPair->second;
+            Node *n = mClient->nodebyhandle(transfer->getType() == MegaTransfer::TYPE_UPLOAD ? transfer->getParentHandle() : transfer->getNodeHandle());
+            while (n)
+            {
+                if (n && n->localnode && n->localnode->sync)
+                {
+                    syncTag = n->localnode->sync->tag;
+                    mTransferToSyncMap[transfer->getTag()] = syncTag;
+                    break;
+                }
+                LOG_warn << "Heartbeat could not get sync tag direclty from transfer handle. Going up";
+                n = n->parent;
+            }
         }
-        else
+
+        if (syncTag)
         {
-            //create new HeartBeatSyncInfo
-            return mHeartBeatedSyncs.insert(std::make_pair(syncTag, std::make_shared<HeartBeatSyncInfo>(syncTag, UNDEF))).first->second;
+            auto hBPair = mHeartBeatedSyncs.find(syncTag);
+            if (hBPair != mHeartBeatedSyncs.end()) //transfer info arrived before having created the HeartBeatSyncInfo object
+            {
+                return hBPair->second;
+            }
+            else
+            {
+                //create new HeartBeatSyncInfo
+                return mHeartBeatedSyncs.insert(std::make_pair(syncTag, std::make_shared<HeartBeatSyncInfo>(syncTag, UNDEF))).first->second;
+            }
```

Comment:
`if (!syncTag)` at this point, it would be an error that we may want to catch. Perhaps an `assert(syncTag)` before the `if (syncTag)`?